src/compiler/parser.ts

```diff
@@ -2143,16 +2166,31 @@ module ts {
             }
         }
 
+        function setContextParameters(explicitFlags: ParserContextFlags = ParserContextFlags.None, ambientFlags: ParserContextFlags = ParserContextFlags.None) {
+            var savedFlags = contextFlags;
+            var generatedFlags = contextFlags & ParserContextFlags.ContextParameterFlags;
+            contextFlags &= ~ParserContextFlags.ContextParameterFlags;
+            contextFlags |= explicitFlags;
+            contextFlags |= generatedFlags & ambientFlags;
+            return savedFlags;
+        }
+
+        function setContextParametersAnd<T>(callback: () => T, explicitFlags?: ParserContextFlags, ambientFlags?: ParserContextFlags): T {
+            var savedFlags = setContextParameters(explicitFlags, ambientFlags);
+            var result = callback();
+            contextFlags = savedFlags;
+            return result;
+        }
+
         function parseParameter(): ParameterDeclaration {
             var node = <ParameterDeclaration>createNode(SyntaxKind.Parameter);
             setModifiers(node, parseModifiers());
             node.dotDotDotToken = parseOptionalToken(SyntaxKind.DotDotDotToken);
 
-            // SingleNameBinding[Yield,GeneratorParameter] : See 13.2.3
-            //      [+GeneratorParameter]BindingIdentifier[Yield]Initializer[In]opt
-            //      [~GeneratorParameter]BindingIdentifier[?Yield]Initializer[In, ?Yield]opt
+            // FormalParameter[Yield,GeneratorParameter,Await,AsyncParameter] : (Modified) See 14.1
+            //      BindingElement[?Yield,?GeneratorParameter,?Await,?AsyncParameter]
 
-            node.name = inGeneratorParameterContext() ? doInYieldContext(parseIdentifierOrPattern) : parseIdentifierOrPattern();
+            node.name = parseIdentifierOrPattern(/*inBindingElement*/ true);

```

Comment:
Not when parsing the identifier or pattern, only the initializers. The grammar can be hard to read.

Here is the grammar (without async/await):
```
BindingElement[Yield,GeneratorParameter] : 
    [+GeneratorParameter] BindingPattern[?Yield,GeneratorParameter] Initializer[In]opt
    [~GeneratorParameter] BindingPattern[?Yield] Initializer[In,?Yield]opt

SingleNameBinding[Yield,GeneratorParameter] : 
    [+GeneratorParameter] BindingIdentifier[Yield] Initializer[In]opt
    [~GeneratorParameter] BindingIdentifier[?Yield] Initializer[In,?Yield]opt
```

Here we are parsing either a binding pattern or an identifier. The parameter qualifier (e.g. `[+GeneratorParameter]`) only really affects the parameters passed to the `Initializer` non-terminal. The `BindingElement` production could have been rewritten as:
```
BindingElement[Yield,GeneratorParameter] : 
    [+GeneratorParameter] BindingPattern[?Yield,?GeneratorParameter] Initializer[In]opt
    [~GeneratorParameter] BindingPattern[?Yield,?GeneratorParameter] Initializer[In,?Yield]opt
```

The `SingleNameBinding` production is a bit odd, in that if `[GeneratorParameter]` is set, it explicitly sets `[Yield]`. If you follow the grammar through, there really should never be a case where `[Yield]` is not set when `[GeneratorParameter]` is set. As such, `SingleNameBinding` could be rewritten as"
```
SingleNameBinding[Yield,GeneratorParameter] : 
    [+GeneratorParameter] BindingIdentifier[?Yield] Initializer[In]opt
    [~GeneratorParameter] BindingIdentifier[?Yield] Initializer[In,?Yield]opt
```
With these productions, the `[GeneratorParameter]` production only ever effects how `Initializer` is parsed.